platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,762 @@
+            // we need to supply another value if we want to use `msg` later to reply.
+            msg.body = sdmmc::Command::default();
+            if let Err(error) = reply.reply_konly(msg.reply_with(response)).await {
+                tracing::warn!(?error, "client hung up...");
+            }
+        }
+    }
+
+    #[tracing::instrument(level = tracing::Level::DEBUG, skip(self, params))]
+    async fn command(&self, params: sdmmc::Command) -> Result<sdmmc::Response, sdmmc::Error> {
+        if self.smhc.smhc_cmd.read().cmd_load().bit_is_set() {
+            return Err(sdmmc::Error::Busy);
+        }
+
+        let cmd_idx = params.index;
+        // TODO: naive `auto_stop` selection, probably only works in case of SD memory cards.
+        // Should this (auto_stop select) be part of the params?
+        let (data_trans, trans_dir, auto_stop) = match params.kind {
+            sdmmc::CommandKind::Control => (false, false, false),
+            sdmmc::CommandKind::Read(len) => (true, false, len > 512),
+            sdmmc::CommandKind::Write(len) => (true, true, len > 512),
+        };
+        let chk_resp_crc = params.rsp_crc;
+        let long_resp = params.rsp_type == sdmmc::ResponseType::Long;
+        let resp_rcv = params.rsp_type != sdmmc::ResponseType::None;
+
+        // Do any required configuration
+        match params.options {
+            sdmmc::HardwareOptions::None => (),
+            sdmmc::HardwareOptions::SetBusWidth(width) => {
+                self.smhc.smhc_ctype.write(|w| match width {
+                    sdmmc::BusWidth::Single => w.card_wid().b1(),
+                    sdmmc::BusWidth::Quad => w.card_wid().b4(),
+                    sdmmc::BusWidth::Octo => w.card_wid().b8(),
+                });
+            }
+        }
+
+        let mut guard = self.isr.lock(self.smhc);
+        guard.data.op = match (params.kind, params.buffer) {
+            (sdmmc::CommandKind::Control, _) => SmhcOp::Control,
+            (sdmmc::CommandKind::Read(cnt), Some(buf)) => SmhcOp::Read {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            (sdmmc::CommandKind::Write(cnt), Some(buf)) => SmhcOp::Write {
+                buf,
+                cnt,
+                auto_stop,
+            },
+            _ => {
+                tracing::error!("did not provide a buffer for read/write");
+                return Err(sdmmc::Error::Other);
+            }
+        };
+
+        // Statically allocate space for 16 DMA descriptors.
+        // Each descriptor can do a transfer of 4KB, giving a max total transfer of 64KB.
+        // By declaring them in this scope the descriptor memory will live long enough,
+        // however currently this is up to the user to guarantuee.
+        // Safety: a zero'ed descriptor is valid and will simply be ignored by the IDMAC.
+        let mut descriptors: [idmac::Descriptor; 16] = unsafe { core::mem::zeroed() };
```

Comment:
I'd like to improve this, but don't know if dynamically allocating memory for descriptors is the way to go.

Reply:
I'm going  to look at the regular DMAC code and try to do things the same way for this internal DMAC.

Reply:
Hey @hawkw, could you point me in the right direction regarding the way descriptors are allocated for the regular DMAC? I don't find references to `mnemos_bitpool` in the code and from what I can see the two current DMAC users (uart + spim) have enough with 1 descriptor per transfer (whereas here the number of required descriptors depends on the size of the transfer).

Reply:
The code I was referring to is the DMAC channel allocator in [`d1_core::dmac`](https://github.com/tosc-rs/mnemos/blob/main/platforms/allwinner-d1/d1-core/src/dmac/mod.rs), particularly the [`Dmac::claim_channel`](https://mnemos.dev/doc/mnemos_d1_core/dmac/struct.dmac#method.claim_channel) method and friends. I think a somewhat similar pattern could be used here, potentially.

I probably wouldn't worry about that right now though. If the current solution works, this might be worth saving for a subsequent PR.